specs/spec.md

```diff
@@ -0,0 +1,319 @@
+# @TODO
+- [ ] Link to parsers.
+- [ ] attestations
+
+# Hobbits
+
+*Written by Dean Eigenmann <dean@status.im> & Rene Nayman <rene@whiteblock.io>* 
+
+## Table of Contents
+
+1. [Abstract](#abstract)
+2. [Flow](#flow)
+3. [Wire Protocol](#wire-protocol)
+    3.1 [Messages](#messages)
+4. [Transport](#transport)
+5. [Protocols](#protocols)
+    5.1 [RPC](#rpc)
+    5.2 [Gossip](#gossip)
+    5.3 [PING](#ping)
+6. [Implementations](#implementations)
+
+## Abstract
+
+Hobbits is a modular wire protocol which allows implementers to experiment with various application logic over a practical network in an agnostic and expedient manner.
+
+Hobbits wire protocol was developed so that Eth2.0 clients could begin network-level testing without requiring libp2p.
+
+## Flow
+
+We layout a network where the nodes are peered in the following manner:
+
+![nodes](./nodes.png)
+
+The following flow dictates how blocks are shared amongst peers:
+
+![sequence diagram](./diagram.png)
+
+1. Node 1 receives a new block `OxCAB` and gossips it to Node 2 and Node 3
+2. Node 2 requests the block bodies for block `OxCAB` from Node 1 in an RPC Command query
+3. Node 1 responds to Node 2 with the block bodies for block `OxCAB` in an RPC Command response
+4. Node 2 receives a new block `OxCAB` and gossips it to Node 1 and Node 3
+
+## Wire Protocol
+
+### Messages
+
+Messages are sent as `utf-8 byte encoded arrays`. The `message` format looks as follows:
+
+```
+EWP <version> <protocol> <header-length> <body-length>
+<header><body>
+```
+
+**NOTE:** Message`body`s along with `header` fields are BSON encoded.
+
+#### Fields
+
+Every hobbit message contains the following fields: 
+
+| Field | Definition | Validity |
+|:------:|----------|:----:|
+| `version` | Defines the EWP version number e.g. `0.2`. | `(\d+\.)(\d+)` |
+| `protocol` | Defines the [protocol](#protocol). | `(RPC\|GOSSIP\|PING)` |
```

Comment:
Are these enum values represented as int or literately "RPC" as string.

Reply:
This link is broken.

Reply:
I would assume as numbers...

Reply:
nah, strings.